packages/@ama-terasu/cli/src/helpers/module.helper.ts

```diff
@@ -0,0 +1,151 @@
+        resolve(dynamicDependenciesPath, 'node_modules'),
+        ...module.paths
+      ]
+    });
+
+    return resolvedPath || null;
+  } catch {
+    return null;
+  }
+};
+
+
+/**
+ * Get the module simplified name
+ *
+ * @param pck package to get name from
+ */
+export const getSimplifiedName = (pck: (PackageJson | SearchResult) & { name: string }) => {
+  return /(?:@[^/]+[/])?(?:amaterasu-)?(.*)/.exec(pck.name)?.[1] || pck.name;
+};
+
+/**
+ * Retrieve the list of modules registered to Amaterasu CLI
+ *
+ * @param dependencies list of the CLI dependencies
+ * @param options Options for the module resolution
+ * @param options.localOnly Resolve module locally only
+ * @returns list of modules to load
+ */
+export const getCliModules = async (dependencies: Record<string, string>, options: { localOnly: boolean } = { localOnly: false }): Promise<ModuleDiscoveryList> => {
+  const npmModulesPromise = options.localOnly ? Promise.resolve({ stdout: '[]'}) : promisify(exec)(`npm search ${moduleKeyword} --json`);
+  const explicitModules = Object.keys(dependencies)
+    .map((moduleName) => getDepPackage(moduleName))
+    .filter((depModule): depModule is PackageJson => {
+      const { keywords } = depModule || { keywords: undefined };
+      return !!keywords && keywords.includes(moduleKeyword);
+    });
+
+  let npmModules: SearchResult[] = [];
+  try {
+    npmModules = JSON.parse((await npmModulesPromise).stdout) as SearchResult[];
+  } catch {
+    console.warn('Failed to execute `npm search`, will contains only installed packages');
+  }
+
+  const map = new Map<string, ModuleDiscovery & { name: string }>();
+  [...npmModules, ...explicitModules]
+    .filter((mod): mod is typeof mod & { name: string } => !!mod.name)
+    .forEach((mod) => map.set(mod.name, Object.assign(mod, { resolutionPath: getInstalledPath(mod), moduleName: getSimplifiedName(mod) })));
+  return [
+    ...map.values()
+  ];
+};
+
+/**
+ * Determine if the package is installed
+ *
+ * @param pck package to get name from
+ */
+export const isInstalled = (pck: ModuleDiscovery): pck is ModuleDiscovery & { resolutionPath: string } => {
+  return !!pck.resolutionPath;
+};
+
+/**
+ * Formatted description
+ *
+ * @param pck package to get name from
+ */
+export const getFormattedDescription = (pck: ModuleDiscovery): string => {
+  return (isInstalled(pck) ? '' : `${chalk.grey.italic('(remote)')} `) + (pck.description || '<Missing description>');
+};
+
+/**
+ * Install a specific package. Will be ignored if the package is already installed
+ *
+ * @param pck Package to install
+ */
+export const installDependency = async (pck: ModuleDiscovery) => {
+  if (!isInstalled(pck)) {
+    if (!existsSync(resolve(dynamicDependenciesPath, 'package.json'))) {
+      if (!existsSync(resolve(dynamicDependenciesPath))) {
+        await fs.mkdir(dynamicDependenciesPath, { recursive: true });
+      }
+      await promisify(exec)('npm init --yes', { cwd: dynamicDependenciesPath });
```

Comment:
It not exactly in the `node_modules`, it creates a folder 'dyn_modules' beside of the CLI's package.json.

> What happens when the version of the cli is upgraded? Do we loose all the installed modules?
When we upgrade the version of the CLI, the dynamic installed package are removed (it is quite safer for now because there is not version constrain validation on installed module so far).

> Does it work with pnp?
I don't think it works with yarn pnp, it is working in this monorepo because the executable is inside the `dist` folder but I am quite sure it will not be the same if it would de as pnp dep.
I don't know if the `yarn dlx` command is based on pnp per default but this CLI is explicitly made to support `npx` (which will lead to global npm install) make it work with yarn pnp would be great but not really mandatory so far.
I will create a task for it.

> What happens if the installed module is outdated or if the user wants a specific version?
There is clearly no version constraint check or upgrade mechanism and it is a miss. I create a task to add it later (if it is OK to not be part of this PR)

Reply:
> It is not exactly in the node_modules, it creates a folder 'dyn_modules' beside of the CLI's package.json.

Isn't the `dyn_modules` folder inside the node_modules installed for `amaterasu/cli`?
Ex: `node_modules/amaterasu/cli/dyn_modules`?
That's what I meant by "a folder inside the node modules folder of amaterasu/cli"

Is it made clear in the doc that the CLI should only be installed globally?

Reply:
Oooooh I think I start to understand your point.
So yes it will be included in parent node_modules (as you mentioned `..../node_modules/amaterasu/cli/dyn_modules`).
Is it the point you wanted to raise and suggested to install it in a different location (outside from the node_module)?

Actually it is not mentioned in the documentation that it should be installed only globally because it is not really the case. Npx is doing a local installation (in a temporary folder) as sub dependency of a package it generates to be able to run the exposed .bin. The CLI is technically working in local installation (in non-pnp context for now, I will try with [preferredUnplugged](https://yarnpkg.com/configuration/manifest#preferUnplugged) solution).
The documentation guides to use `npx` but may not be enough clear on the limitation indeed.

Reply:
I have done some tests on `yarn dlx` and it is ok when using `nodeLinker: node_modules` configuration.
The issue that is still remain is for non-node_modules strategies, I add a warning in the documentation regarding this restriction

Reply:
> Is it the point you wanted to raise and suggested to install it in a different location (outside from the node_module)?

No not necessarily, I just wanted to make sure I understood how it worked otherwise the comment about an update of the CLI version removing all remote deps would have been invalid.


I guess the `which will lead to global npm install` confused me 😸 
So just to summarize:
We don't recommend to add it as a dev dep of a project nor to install it globally, it should be installed and run independently via `npx` preferably to avoid conflicts with yarn pnp.